docs/failure-detector/failure-detector.md

```diff
@@ -49,4 +49,76 @@ In the example:
  - node `a` is a decision maker, it has 3 successful connections
  - node `c` is a failed node, it has same number of successful connections like node `b` but is has higher node name.
  - node `a` will add node `c` to the unresponsive list.     
+ 
+### Implementation
+
+**Remote Monitoring Service** constitutes of failure and healing monitoring and handling.
+This service is responsible for heartbeat and aggregating the cluster view. 
+This is updated in the shared context with the management server which serves the heartbeat responses.
+The failure detector updates the unreachable nodes in the layout.
+The healing detector heals nodes which were previously marked unresponsive but have now healed.
+
+
+**Collect poll report**
+Each poll round consists of iterations. In each iteration, the FailureDetector pings all responsive nodes 
+in the layout and also collects their node states to provide cluster state.
+To provide more resilient failre detection mechanism, the failure detector collects few "poll reports" 
+equal to `failureThreshold` number then aggregates final poll report
+
+Once poll report collected, the `failureDetectionTask` is triggered by Remote Monitoring Service.
+To detect a failure in the cluster the Remote Monitoring Service gets `cluster state` from a poll report.
+The cluster state contains information of the cluster.
+
+**ClusterAdvisor** detects failures using cluster state. There could be different types of cluster advisors.
+Currently we use `CompleteGraphAdvisor` - any alive node must be connected to all other alive nodes in the cluster.
+`CompleteGraphAdvisor`:
+ - The advisor transforms cluster state into the cluster graph and makes it symmetric.
+ It means converting a cluster graph which could have asymmetric failures to a graph with symmetric failures between nodes.
+ For instance: <br/>
+ {a: [{a: OK, b: FAILED}]} <br/>
+ {b: [{a: OK, b: OK}]} <br/>
+ Node A believes that node B is disconnected <br/>
+ Node B believes that node A is connected <br/>
+ The graph will be changed to: <br/>
+ {a: [{a: OK, b: FAILED}]} <br/>
+ {b: [{a: FAILED, b: OK}]} <br/>
+ Node B is not connected to node A anymore.
+ 
+**Decision maker** - is a node which can decide to add a failed node into the unresponsive list.
+The decision maker is an technique, used to reduce parallel updates of the cluster state.
+Still, each node can choose its own decision maker.
```

Comment:
Why "each node can choose its own decision marker"? Does it mean each node choose the decision maker based on its own cluster view such that the decision maker meet all the requirements list below?

Reply:
Also, what if different decision markers have different decisions on failed nodes?

Reply:
1. Yes you are right each node choose the decision maker based on its own cluster view 
2. It's possible that two different nodes decide to be a decision maker at the same time and can have different decisions on failed nodes. In that case, nodes would be added into the unresponsive list. On the next iteration, nodes would have a more consistent view on the cluster state and a node accidentally added to the unresponsive list will be healed

Reply:
1. Yes you are right each node choose the decision maker based on its own cluster view

2. It's possible that two different nodes decide to be a decision maker at the same time and can have different decisions on failed nodes. In that case, nodes would be added into the unresponsive list. On the next iteration, nodes would have a more consistent view on the cluster state and a node accidentally added to the unresponsive list will be healed